### ISSUE TITLE
self-attribution score short cut

### ISSUE THREAD

**Issue by TheaperDeng** [TRAIS-Lab/dattri#160]
We need a short cut to calculate self-importance or self-attribution score.

Currently, we calculate a N_train * N_train matrix while a large portion of them (other than the main diag) are wasted.

**Comment by lalalalayula** [comment]
We can just calculate the dot product of train_batch_grad  and inverse hessian matrix to get the self-attribution score matrix.

**Comment by TheaperDeng** [comment]
> We can just calculate the dot product of train_batch_grad and inverse hessian matrix to get the self-attribution score matrix.

Yes, generally it should be. Please consider two implementation tricks
- It should be $$g^TH^-1g$$, where g is train gradient of a training sample
- No need to calculate the "cross influence" between different training samples, that's the key reason why we need a short cut.

**Comment by yiliey** [comment]
1.What methods (external methods, i.e., the ones that will be directly called by users) will I add/modify. Use docstring directly to answer the question.
I will simply change the input, output and the implementation of the attribute function for every algorithm. I will take the one of influence funtion as an example.
```python
def attribute(self, train_dataloader: DataLoader) -> torch.Tensor:

        if self.full_train_dataloader is None:
            self.full_train_dataloader = train_dataloader
        _check_shuffle(train_dataloader)
        diagonal_output = torch.zeros(
            size=(len(train_dataloader.dataset), ),
            device=self.device,
        )
        for train_batch_idx, (train_inputs, train_targets) in enumerate(train_dataloader):
            train_batch_data = (train_inputs.to(self.device), train_targets.to(self.device))
            train_batch_grad = self.grad_func(self.params, train_batch_data)
            # for test_batch_idx, (test_inputs, test_targets) in enumerate(test_dataloader):
            #     test_batch_data = (test_inputs.to(self.device), test_targets.to(self.device))
            #     test_batch_grad = self.grad_func(self.params, test_batch_data)

            ihvp = 0
            for full_inputs, full_targets in self.full_train_dataloader:
                full_data = (full_inputs.to(self.device), full_targets.to(self.device))
                self.ihvp_func = self.ihvp_solver(partial(self.grad_func, params=self.params))
                ihvp += self.ihvp_func(self.params, train_batch_grad).detach()
                row_st = train_batch_idx * train_dataloader.batch_size
                row_ed = min(row_st + train_dataloader.batch_size, len(train_dataloader.dataset))
            diagonal_output[row_st:row_ed] += torch.mul(train_batch_grad, ihvp)
        
        return diagonal_output
```
2.How will users use this function? Please use a very simple (but comprehensive) example call to answer the question.
```python
task = AttributionTask(loss_func=loss,
                       model=self.model,
                       checkpoints=model.state_dict())
influence_calc = InfluenceFunction(model, device="cpu")
    influence_matrix = influence_calc.attribute(train_dataloader)
```
3.How will I implement it?
Change the original influence matrix into a one-dimension tensor. Also, changing the dot product of gradient and transpose of ihvp into the elementwise product of  gradient and  ihvp to directly get the influence tensor.
4.Design Adjustment
Now, I still fail to integrate the new method into the whole project, but I can implement it in the seperate test file.

**Comment by TheaperDeng** [comment]
1. Please don't call it `attribute`. It will be the same as the original `attribute` method (which should be kept). something like `self_attribute` would be good
2. Please add a docstring like https://github.com/TRAIS-Lab/dattri/blob/02aa0906388d5dec420e7366cfffd7ab77a94e69/dattri/algorithm/influence_function.py#L119-L133 here to show the input and output formal definition
3. Please refer to

- https://github.com/TRAIS-Lab/dattri/blob/02aa0906388d5dec420e7366cfffd7ab77a94e69/dattri/algorithm/base.py#L258-L262 implementation for your `self_attribute` method in influence function
- https://github.com/TRAIS-Lab/dattri/blob/02aa0906388d5dec420e7366cfffd7ab77a94e69/dattri/algorithm/trak.py#L187-L191 implementation for your `self_attribute` method in trak
- https://github.com/TRAIS-Lab/dattri/blob/02aa0906388d5dec420e7366cfffd7ab77a94e69/dattri/algorithm/tracin.py#L71-L75  implementation for your `self_attribute` method in tracin

4. There is no need to change this class https://github.com/TRAIS-Lab/dattri/blob/02aa0906388d5dec420e7366cfffd7ab77a94e69/dattri/algorithm/influence_function.py#L55-L222
